text/0024-track-validator-set-when-synchronizing.org

```diff
@@ -0,0 +1,142 @@
+for each era is essential to securely joining the network.
+
+** Guide-level explanation
+:PROPERTIES:
+:CUSTOM_ID: guide-level-explanation
+:END:
+
+Joining a network requires that a node process one block at a time
+(under the current scheme). It does so until it becomes current. The
+node then runs the /highway/ protocol.
+
+The following depicts the process a joining node will take assuming an
+era height of 2.  Checking will not be implemented until
+[[https://github.com/casperlabs/ceps/pull/1234][casperlabs/ceps#1234]].
+
+#+BEGIN_SRC svgbob :file images/0024/joining.svg :exports results
+
+            Validators    Blocks       Joining Node Actions
+
+       __
+       \/   .~~~~~~~.     ,-------.
+       __   : A B C :~~~~/ Era 1 /<--- Initialize validator set
+       \/   `~~~~~~~'   '----+--'
+       __                    |
+ -+-   \/                   [#]<------ Check A/B/C sigs
+  |    __                    |
+       \/                   [#]<------ Check A/B/C sigs
+  |    __                    |
+  |    \/     .~~~~~.     ,--+----.
+       __     : A B :~~~~/ Era 2 /<--- Update validator set
+ |\/|  \/     `~~~~~'   '----+--'
+ |  |  __                    |
+       \/                   [#]<------ Check A/B sigs
+ +--   __                    |
+ |--   \/                   [#]<------ Check A/B sigs
+ +--   __                    |
+       \/   .~~~~~~~.     ,--+----.
+       __   : A B D :~~~~/ Era 3 /<--- Update validator set
+       \/   `~~~~~~~'   '----+--'
+       __                    |
+       \/                   [#]<------ Check A/B/D sigs
+       __                    :
+       \/                    :
+
+
+  +--------+-------------------------------------+
+  | Legend |                                     |
+  +--------+                                     |
+  |                                              |
+  |                          |                   |
+  |  .~~~~.                 ,+--.                |
+  |  :    : Validator Set  /   /   Switch Block  |
+  |  `~~~~'               '--+'                  |
+  |                          |                   |
+  |                                              |
+  |                          |                   |
+  |                         [#]    Normal Block  |
+  |                          |                   |
+  |                                              |
+  +----------------------------------------------+
+
+
+#+END_SRC
+
+#+RESULTS:
+[[file:images/0024/joining.svg]]
+
+As we can see above, a switch block can change the validator
+set. Joining nodes needs to track this.
+
+** Reference-level explanation
+:PROPERTIES:
+:CUSTOM_ID: reference-level-explanation
+:END:
+
+/Synchronization mode/ is mediated by the variant
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain_sync.rs#L85-L93][=linear_chain_sync::State::SyncingDescendants]]=.
```

Comment:
This renders poorly on GitHub:
![image](https://user-images.githubusercontent.com/5120801/101690244-86106600-3a6d-11eb-8787-c61aaf8ae2c6.png)

Reply:
Should be fixed here: https://github.com/CasperLabs/ceps/pull/24/commits/e440c640aaefb9589df2eee99259e06a0c05c2f6